From 4a60d72512210fb09168461dc2e915d88a431d62 Mon Sep 17 00:00:00 2001
From: Haseeb Ashraf <haseeb.ashraf@siemens.com>
Date: Thu, 22 Oct 2026 14:25:16 +0000
Subject: [PATCH] host: remove hard dependency on X11

Build the host backend without X11 on Linux:

 - include <GL/glx.h> only when the glx meson option is enabled
   (-DENABLE_GLX), so that -Dglx=disabled builds without X11 headers
 - use VK_USE_PLATFORM_WAYLAND_KHR instead of VK_USE_PLATFORM_XCB_KHR
 - replace the X11 sub-window implementation by a no-op one; the
   embedder does not use native sub-windows

Upstream-Status: Inappropriate [custom implementation]

Signed-off-by: Haseeb Ashraf <haseeb.ashraf@siemens.com>
---
 host/gl/glestranslator/common/gl_dispatch.cpp |  2 ++
 host/gl/glestranslator/common/meson.build     |  4 +++
 host/meson.build                              |  2 +-
 host/native_window/meson.build                |  2 +-
 .../native_sub_window_wayland.cpp             | 30 +++++++++++++++++++
 host/vulkan/cereal/meson.build                |  2 +-
 host/vulkan/emulated_textures/meson.build     |  2 +-
 host/vulkan/meson.build                       |  2 +-
 8 files changed, 41 insertions(+), 5 deletions(-)
 create mode 100644 host/native_window/native_sub_window_wayland.cpp

diff --git a/host/gl/glestranslator/common/gl_dispatch.cpp b/host/gl/glestranslator/common/gl_dispatch.cpp
index f2b9f42..47ca23a 100644
--- a/host/gl/glestranslator/common/gl_dispatch.cpp
+++ b/host/gl/glestranslator/common/gl_dispatch.cpp
@@ -22,7 +22,9 @@
 #include "gfxstream/common/logging.h"
 
 #ifdef __linux__
+#ifdef ENABLE_GLX
 #include <GL/glx.h>
+#endif
 #elif defined(WIN32)
 #include <windows.h>
 #endif
diff --git a/host/gl/glestranslator/common/meson.build b/host/gl/glestranslator/common/meson.build
index c838d6e..407dfa6 100644
--- a/host/gl/glestranslator/common/meson.build
+++ b/host/gl/glestranslator/common/meson.build
@@ -1,6 +1,10 @@
 # Copyright 2023 Android Open Source Project
 # SPDX-License-Identifier: Apache-2.0
 
+if get_option('glx') != 'disabled'
+  gfxstream_host_args += ['-DENABLE_GLX']
+endif
+
 files_lib_gl_common = files(
   'rgtc.cpp',
   'framebuffer_data.cpp',
diff --git a/host/meson.build b/host/meson.build
index 1e8875a..9e22f7f 100644
--- a/host/meson.build
+++ b/host/meson.build
@@ -140,7 +140,7 @@ elif host_machine.system() == 'darwin'
   gfxstream_backend_cpp_args += '-DVK_USE_PLATFORM_METAL_EXT'
   gfxstream_backend_cpp_args += '-DVK_USE_PLATFORM_MACOS_MVK'
 elif host_machine.system() == 'linux'
-  gfxstream_backend_cpp_args += '-DVK_USE_PLATFORM_XCB_KHR'
+  gfxstream_backend_cpp_args += '-DVK_USE_PLATFORM_WAYLAND_KHR'
 endif
 
 deps_gfxstream_backend = []
diff --git a/host/native_window/meson.build b/host/native_window/meson.build
index fe6a556..b6f8e41 100644
--- a/host/native_window/meson.build
+++ b/host/native_window/meson.build
@@ -23,7 +23,7 @@ elif host_machine.system() == 'windows'
   files_lib_host_native_window += files('native_sub_window_win32.cpp')
   includes_lib_host_native_window += [inc_opengl_headers]
 elif host_machine.system() == 'linux' and use_gles
-  files_lib_host_native_window += files('native_sub_window_x11.cpp')
+  files_lib_host_native_window += files('native_sub_window_wayland.cpp')
   includes_lib_host_native_window += [inc_opengl_headers]
 elif host_machine.system() == 'qnx'
   files_lib_host_native_window += files(
diff --git a/host/native_window/native_sub_window_wayland.cpp b/host/native_window/native_sub_window_wayland.cpp
new file mode 100644
index 0000000..17228b8
--- /dev/null
+++ b/host/native_window/native_sub_window_wayland.cpp
@@ -0,0 +1,30 @@
+#include "gfxstream/host/native_sub_window.h"
+
+EGLNativeWindowType createSubWindow(FBNativeWindowType,
+                                    int,
+                                    int,
+                                    int,
+                                    int,
+                                    float,
+                                    SubWindowRepaintCallback,
+                                    void*,
+                                    int) {
+    return 0;
+}
+
+void destroySubWindow(EGLNativeWindowType) {
+}
+
+int moveSubWindow(FBNativeWindowType,
+                  EGLNativeWindowType,
+                  int,
+                  int,
+                  int,
+                  int,
+                  float) {
+    return 0;
+}
+
+void* getNativeDisplay() {
+    return nullptr;
+}
diff --git a/host/vulkan/cereal/meson.build b/host/vulkan/cereal/meson.build
index d1c2b82..da73d2a 100644
--- a/host/vulkan/cereal/meson.build
+++ b/host/vulkan/cereal/meson.build
@@ -24,7 +24,7 @@ elif host_machine.system() == 'darwin'
   cereal_cpp_args += '-DVK_USE_PLATFORM_METAL_EXT'
   cereal_cpp_args += '-DVK_USE_PLATFORM_MACOS_MVK'
 elif host_machine.system() == 'linux'
-  cereal_cpp_args += '-DVK_USE_PLATFORM_XCB_KHR'
+  cereal_cpp_args += '-DVK_USE_PLATFORM_WAYLAND_KHR'
 endif
 
 lib_vulkan_cereal = static_library(
diff --git a/host/vulkan/emulated_textures/meson.build b/host/vulkan/emulated_textures/meson.build
index 56eed17..1728c57 100644
--- a/host/vulkan/emulated_textures/meson.build
+++ b/host/vulkan/emulated_textures/meson.build
@@ -17,7 +17,7 @@ elif host_machine.system() == 'darwin'
   emulated_texture_args += '-DVK_USE_PLATFORM_METAL_EXT'
   emulated_texture_args += '-DVK_USE_PLATFORM_MACOS_MVK'
 elif host_machine.system() == 'linux'
-  emulated_texture_args += '-DVK_USE_PLATFORM_XCB_KHR'
+  emulated_texture_args += '-DVK_USE_PLATFORM_WAYLAND_KHR'
 endif
 
 lib_emulated_textures = static_library(
diff --git a/host/vulkan/meson.build b/host/vulkan/meson.build
index 602458e..f45e339 100644
--- a/host/vulkan/meson.build
+++ b/host/vulkan/meson.build
@@ -63,7 +63,7 @@ elif host_machine.system() == 'windows'
 elif host_machine.system() == 'qnx'
   vulkan_server_cpp_args += '-DVK_USE_PLATFORM_SCREEN_QNX'
 elif host_machine.system() == 'linux'
-  vulkan_server_cpp_args += '-DVK_USE_PLATFORM_XCB_KHR'
+  vulkan_server_cpp_args += '-DVK_USE_PLATFORM_WAYLAND_KHR'
 endif
 
 lib_vulkan_server = static_library(
